### PR TITLE
Only discover complex types when configured through pre-convention model configuration or annotation

### LINKED ISSUE #31344: Only discover complex types when configured through pre-convention model configuration
In the current bits, when a CLR type is configured as a complex type in one place, it's picked up as a complex type by convention in other places as well; so a CLR type can be used as a complex type in many places in the model, but only explicitly configured as a complex type once. This means that if the one explicit configuration is removed for some reason, all others become broken in various ways.

Instead, we can add an API to pre-convention model configuration to configure a CLR type as complex, and otherwise require that users explicitly configure.

Note the interaction with #31275, which is about automatically configuring *nested* complex types as complex. Implementing the behavior above removes the problematic part raised in https://github.com/dotnet/efcore/issues/31275#issuecomment-1642938765.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Throw for complex properties of value types
Throw for optional complex properties

Part of #13947
Fixes #31344

### CODE REVIEW

File: src/EFCore/Properties/CoreStrings.Designer.cs [hunk 1]
```diff
@@ -480,6 +480,14 @@ public static string ComplexPropertyIndexer(object? type, object? property)
                 GetString("ComplexPropertyIndexer", nameof(type), nameof(property)),
                 type, property);
 
+        /// <summary>
+        ///     Configuring the complex property '{type}.{property}' as optional is not supported, call 'IsRequired()'.
```
Comment (roji) [inline src/EFCore/Properties/CoreStrings.Designer.cs:484]: Maybe point to #31376 to get votes/feedback?

File: test/EFCore.Tests/ModelBuilding/ComplexTypeTestBase.cs [hunk 2]
```diff
@@ -1518,10 +1518,10 @@ public virtual void Complex_properties_discovered_by_convention()
                 .Entity<ComplexProperties>()
                 .ComplexProperty(e => e.Customer);
 
-            modelBuilder
-                .Entity<ValueComplexProperties>()
-                .Ignore(e => e.Tuple)
-                .ComplexProperty(e => e.Label);
+            //modelBuilder
```
Comment (roji) [inline test/EFCore.Tests/ModelBuilding/ComplexTypeTestBase.cs:1521]: :eyes:

File: src/EFCore.Relational/Metadata/Conventions/SharedTableConvention.cs [hunk 3]
```diff
@@ -269,14 +269,19 @@ protected virtual bool TriggersUniqueAcrossTables
...
                 {
-                    properties[columnName] = property;
-                    properties[newOtherColumnName] = otherProperty;
+                    columns[columnName] = property;
+                    columns[newOtherColumnName] = otherProperty;
                 }
             }
         }
+
+        foreach (var complexPropery in type.GetDeclaredComplexProperties())
```
Comment (roji) [inline src/EFCore.Relational/Metadata/Conventions/SharedTableConvention.cs:281]:
```suggestion
        foreach (var complexProperty in type.GetDeclaredComplexProperties())
```

File: test/EFCore.Tests/ModelBuilding/TestModel.cs [hunk 4]
```diff
@@ -855,19 +855,19 @@ protected interface IReplaceable
     protected class ComplexProperties
     {
         public int Id { get; set; }
-        public Customer? Customer { get; set; }
-        public DoubleProperty? DoubleProperty { get; set; }
-        public IndexedClass? IndexedClass { get; set; }
-        public Quarks? Quarks { get; set; }
+        public Customer Customer { get; set; } = null!;
```
Comment (roji) [inline test/EFCore.Tests/ModelBuilding/TestModel.cs:858]: Can also put required on the properties, the modern way etc.

**Review by roji — APPROVED** [review]
LGTM

File: test/EFCore.Tests/ModelBuilding/ComplexTypeTestBase.cs (hunk 2 above)
Comment (AndriySvyryd) [reply test/EFCore.Tests/ModelBuilding/ComplexTypeTestBase.cs:1521]: Value type support

**Pull request merged**